docs/advanced/cast/custom.rst

```diff
@@ -1,93 +1,135 @@
+              PyErr_Clear();
+              return false;
             }
-        };
-    }} // namespace pybind11::detail
+            return true;
+        }
+        /**
+         * Conversion part 2 (C++ -> Python): convert an inty instance into
+         * a Python object. The second and third arguments are used to
+         * indicate the return value policy and parent object (for
+         * ``return_value_policy::reference_internal``) and are generally
+         * ignored by implicit casters.
+         */
+        static handle cast(inty src, return_value_policy /* policy */, handle /* parent */) {
+            return PyLong_FromLong(src.long_value);
+        }
+    };
+
+    } // namespace user_space
 
 .. note::
 
-    A ``type_caster<T>`` defined with ``PYBIND11_TYPE_CASTER(T, ...)`` requires
+    A caster class using with ``PYBIND11_TYPE_CASTER(T, ...)`` requires
     that ``T`` is default-constructible (``value`` is first default constructed
-    and then ``load()`` assigns to it).
+    and then ``load()`` assigns to it). It is possible but more involved to define
+    a caster class for types that are not default-constructible.
+
+The caster class defined above can be plugged into pybind11 in two ways:
+
+* Starting with pybind11 v2.10, a new — and the recommended — alternative is to *declare* a
+  function named ``pybind11_select_caster``:
+
+  .. code-block:: cpp
+
+    namespace user_space {
+
+    inty_type_caster pybind11_select_caster(inty*);
+
+    } // namespace user_space
+
+  The argument is a *pointer* to the C++ type, the return type is the caster type.
+  This function has no implementation! Its only purpose is to associate the C++ type
+  with its caster class. pybind11 exploits C++ Argument Dependent Lookup
+  (`ADL <https://en.cppreference.com/w/cpp/language/adl>`_)
+  to discover the association.
+
+  Note that ``pybind11_select_caster`` can alternatively be declared as a ``friend``
+  function of the C++ type, if that is practical and preferred:
+
+  .. code-block:: cpp
+
+      struct inty {
+          ...
+          friend inty_type_caster pybind11_select_caster(inty*);
+      }
+
+* An older alternative is to specialize the ``pybind11::detail::type_caster<T>`` template.
+  Although the ``detail`` namespace is involved, adding a ``type_caster`` specialization
+  is explicitly allowed:
+
+  .. code-block:: cpp
+
+      namespace pybind11 { namespace detail {
+          template <> struct type_caster<user_space::inty> : user_space::inty_type_caster {};
+      }} // namespace pybind11::detail
+
+  .. note::
+      ``type_caster` specializations may be full (as in this simple example) or partial.
 
-.. warning::
+  .. warning::
 
-    When using custom type casters, it's important to declare them consistently
-    in every compilation unit of the Python extension module. Otherwise,
-    undefined behavior can ensue.
+      When using this method, it is important to declare the specializations
```

Comment:
Unfortunately, the ADL method does not solve the ODR issue.
For example, for a given `T` there can be only one definition of `load_caster<T>(const handle&)` (defined below), so `make_caster<T>` must be consistent across all modules.